configlib/config.go

```diff
@@ -38,9 +38,61 @@ func NewConfig(cfg *PkgrConfig) {
 		rVersion := rcmd.GetRVersion(&rs)
 		cfg.Library = getLibraryPath(cfg.Lockfile.Type, cfg.RPath, rVersion, rs.Platform, cfg.Library)
 	}
+
+	// For all cfg values that can be paths, make sure that ~ is expanded to the home directory.
+	cfg.Library = expandTilde(cfg.Library)
+	cfg.RPath = expandTilde(cfg.RPath)
+	cfg.Tarballs = expandTildes(cfg.Tarballs)
+	cfg.Repos = expandTildesRepos(cfg.Repos)
+	cfg.Logging.All = expandTilde(cfg.Logging.All)
+	cfg.Logging.Install = expandTilde(cfg.Logging.Install)
+	cfg.Cache = expandTilde(cfg.Cache)
+
 	return
 }
 
+/// expand the ~ at the beginning of a path to the home directory.
+/// consider any problems a fatal error.
+func expandTilde(p string) string {
+	expanded, err := homedir.Expand(p)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"path": p,
+			"error": err,
+		}).Fatal("problem parsing config file -- could not expand path")
+	}
+	return expanded
+}
+
+/// For a list of paths, expand the ~ at the beginning of each path to the home directory.
+/// consider any problems a fatal error.
+func expandTildes(paths []string) []string {
+	var expanded []string
+	for _, p := range paths {
+		newPath := expandTilde(p)
+		expanded = append(expanded, newPath)
+	}
+	return expanded
+}
+
+/// In the PkgrConfig object, Repos are stored as a list of key-value pairs.
+/// Keys are repo names and values are paths to those repos
+/// For each key-value pair, expand the prefix ~ to be the home directory, if applicable.
+/// consider any problems a fatal error.
+func expandTildesRepos(repos []map[string]string) []map[string]string {
+	var expanded []map[string]string
+	//expanded := make(map[string]string)
+	for _, keyValuePair := range repos {
+		kvpExpanded := make(map[string]string)
```

Comment:
whats the reasoning for initializing a new map each time then growing a new object vs just modifying the values in the original map?

Reply:
I'm trying to be consistent with how the Repos structure in PkgrConfig is implemented. IIUC, The Repos structure is a slice of maps, with each individual map being a single key-value pair, so we're replicating that exactly here. That's why this particular function is called `expandTildesRepos` instead of something like `expandTildesMap`

Reply:
What?

I'm saying why create a _new_ map each time. Why not modify the original value in the existing map

repo[key] = homedir.Expand(repo[key])

Reply:
Ah, my bad. 

I don't have an explicit reason for this, but I think instinctively I wanted to treat the Repos object in PkgrConfig as immutable to avoid any bugs that might've come up as a result of modifying by reference.

I'll make the change.

Reply:
Walking back a bit,

Unless you're strongly opinionated on this, I'd actually rather leave it as it is so that all of the "expandTilde" functions work the same way.

```
	cfg.Library = expandTilde(cfg.Library)
	cfg.RPath = expandTilde(cfg.RPath)
	cfg.Tarballs = expandTildes(cfg.Tarballs)
	cfg.Repos = expandTildesRepos(cfg.Repos)
	cfg.Logging.All = expandTilde(cfg.Logging.All)
	cfg.Logging.Install = expandTilde(cfg.Logging.Install)
	cfg.Cache = expandTilde(cfg.Cache)
```
I find the above code nice and consistent, simple to reason about, etc, whereas having just one of those lines suddenly be different
```
	cfg.Library = expandTilde(cfg.Library)
	cfg.RPath = expandTilde(cfg.RPath)
	cfg.Tarballs = expandTildes(cfg.Tarballs)
	expandTildesRepos(cfg.Repos)
	cfg.Logging.All = expandTilde(cfg.Logging.All)
	cfg.Logging.Install = expandTilde(cfg.Logging.Install)
	cfg.Cache = expandTilde(cfg.Cache)
```
Seems kinda strange. I don't think the performance/memory implications in this case are big enough to worry about, either.